components/feature/autofill/src/main/java/mozilla/components/feature/autofill/structure/ParsedStructure.kt

```diff
@@ -24,6 +26,27 @@ internal data class ParsedStructure(
     val packageName: String
 )
 
+/**
+ * Try to find a domain in the [ParsedStructure] for looking up logins. This is either a "web domain"
+ * for web content the third-party app is displaying (e.g. in a WebView) or the package name of the
+ * application transformed into a domain. In any case the [publicSuffixList] will be used to turn
+ * the domain into a "base" domain (public suffix + 1) before returning.
+ */
+internal suspend fun ParsedStructure.getLookupDomain(publicSuffixList: PublicSuffixList): String {
+    val domain = if (webDomain != null && Browsers.isBrowser(packageName)) {
+        // If the application we are auto-filling is a known browser and it provided a webDomain
+        // for the content it is displaying then we try to autofill for that.
+        webDomain
+    } else {
+        // We reverse the package name in the hope that this will resemble a domain name. This is
+        // of course fragile. So we want to find better mechanisms in the future (e.g. looking up
+        // what URLs the application registers intent handlers for).
+        packageName.split('.').asReversed().joinToString(".")
```

Comment:
we'll probably end up with a mapping of top 100X packages to their real domains, and use this as a fallback, won't we...

Reply:
Yeah, possible. With this thing, intent lookup and custom search we may get a good coverage. Even with a custom list (which is a bit painful to maintain) we may need to check digital asset links to make sure we are talking to the right app.